File: combined.js
const fs = require("fs");
const { randomInt } = require("crypto");
const readline = require("readline");

const DIRECTORY_DB = "./db/";
const ACCOUNTS_DB  = "./db/accounts.txt";
const PURCHASES_DB = "./db/purchases.txt";

const rl = readline.createInterface({
    input: process.stdin,
    output: process.stdout
});

const products = {
    1: {
        name        : "Maçã",
        price       : 1.99,
        description : "Vermelha/verde"
    },
    2: {
        name        : "Manga",
        price       : 5.5,
        description : "Rosa/espada"
    },
    3: {
        name        : "Laranja",
        price       : 2.49,
        description : "Lima/pera"
    },
    4: {
        name        : "Banana",
        price       : 0.99,
        description : "Prata/nanica"
    },
    5: {
        name        : "Uva",
        price       : 2.9,
        description : "Tompson/niagara"
    }
};

var cart = [];

function validInput(text) {
    const output = new Promise((resolve) => {
        rl.question(text, (input) => {
            resolve(input.trim());
        });
    });
    
    return output;
}

function parseTXT(filepath) {
    const data = {};
    var lastId = '-1';
    
    const file = fs.readFileSync(filepath, "utf8");
    const content = file.split("\n")
                        .map((line) => line.trim())
                        .filter((line) => line !== "");
    
    content.forEach((line) => {
        const match = line.match(/([a-zA-Z0-9_\-]+)(?:\s+)?(?::)(?:\s+)?(.+)/);
        
        if (match === null) {
            return;
        }
        
        const keyName  = match[1];
        const keyValue = match[2];
        
        if (keyName === "id") {
            lastId = keyValue;
            data[keyValue] = {};
        } else {
            data[lastId][keyName] = keyValue;
        }
    });
    
    return data;
}

function _customPrice(price, dot = ",", money = "R$") {
    const render = price.toFixed(2).replace('.', dot);
    return `${money} ${render}`;
}

function _productExists(productId) {
    const IDs = Object.keys(products);
    
    if (productId in IDs) {
        return true;
    }
    
    return false;
}


function printProducts() {
    for (const productId in products) {
        const product = products[productId];
        
        const name        = product.name;
        const price       = product.price;
        const description = product.description;
        
        console.log(`[${productId}] ${name}: ${_customPrice(price)} - ${description}`);
    }
}

function addToCart(index, quantity) {
    const productId = String(index);
    
    if (!_productExists(productId)) {
        throw new Error()
    }
    
    if (cart[productId] === undefined) {
        cart[productId] = 0;
    }
    
    cart[productId] += quantity;
}

function removeFromCart(index, quantity) {
    const productId = String(index);
    
    if (!_productExists(productId)) {
        throw new Error()
    }
    
    cart[productId] -= quantity;
    
    if (cart[productId] < 1) {
        delete cart[productId];
    }
}

function showCart() {
    if (cart.length === 0) {
        console.log("O carrinho está vazio.");
        return;
    }
    
    var total = 0.0;
    cart.forEach((quantity, productId) => {
        const product = products[productId];
        
        const name        = product.name;
        const price       = product.price * quantity;
        const description = product.description;
        
        total += price;
        console.log(`[${productId}] ${name} (${quantity}): ${_customPrice(price)} - ${description}`);
    });
    
    console.log(`\nPreço Total: ${_customPrice(total)}`);
}

function payProducts(accountId) {
    if (cart.length === 0) {
        console.log("Não há nenhum produto em seu carrinho para ser pago.");
        return;
    }
    
    const purchases = parseTXT(PURCHASES_DB);
    purchases[accountId] = purchases[accountId] ?? {};
    
    var total = 0.0;
    for (const productId in cart) {
        if (purchases[accountId][productId] === undefined) {
            purchases[accountId][productId] = 0;
        } else {
            purchases[accountId][productId] = parseInt(purchases[accountId][productId]);
        }
        
        purchases[accountId][productId] += parseInt(cart[productId]);
        total += products[productId].price * purchases[accountId][productId];
        
        delete cart[productId];
    }
    
    
    fs.unlinkSync(PURCHASES_DB);
    
    for (const userId in purchases) {
        fs.appendFileSync(PURCHASES_DB, `id : ${userId}\n`);
        
        for (const productId in purchases[userId]) {
            const quantity = purchases[userId][productId];
            fs.appendFileSync(PURCHASES_DB, `${productId} : ${quantity}\n`);
        }
    }
    
    
    console.log(
        "Compra feita com sucesso!\n" +
        `Foi pago um total de R$ ${_customPrice(total)}`);
}

function viewPurchases(accountId) {
    const purchases = parseTXT(PURCHASES_DB);
    const personal  = purchases[accountId] ?? {};
    
    if (personal.length === 0) {
        console.log("Não foi feita nenhuma compra nesta conta.");
        return;
    }
    
    var total = 0.0;
    for (const productId in personal) {
        const product  = products[productId];
        const quantity = parseInt(personal[productId]);
        
        const name        = product.name;
        const price       = product.price * quantity;
        const description = product.description;
        
        total += price;
        console.log(`[${productId}] ${name} (${quantity}): ${_customPrice(price)} - ${description}`);
    }
    
    console.log(`\nTotal Pago: ${_customPrice(total)}`);
}

function _getRandomId(limit = 999999999999) {
    const accounts = parseTXT(ACCOUNTS_DB);
    const usedIds = Object.keys(accounts);
    let generatedId;
    
    var loop = true;
    while (loop) {
        generatedId = String(randomInt(1, limit+1));
        
        if (!usedIds.includes(generatedId)) {
            loop = false;
        }
    }
    
    return generatedId;
}

function getAccount(name, password) {
    const accounts = parseTXT(ACCOUNTS_DB);
    var output = [ null, null ];
    
    for (const accountId in accounts) {
        const account = accounts[accountId];
        
        if (account.name.toLowerCase() === name.toLowerCase() &&
            account.password === password) {
            output[0] = accountId;
            output[1] = account;
            break;
        }
    }
    
    return output;
}

function addAccount(name, phoneNumber, dateBirth, cep, password) {
    const account = getAccount(name, password);
    
    if (account[0] !== null && account[1] !== null) {
        return null;
    }
    
    const generatedId = _getRandomId();
    
    fs.appendFileSync(ACCOUNTS_DB,
        `id           : ${generatedId}\n` +
        `name         : ${name}\n` +
        `phone_number : ${phoneNumber}\n` +
        `date_birth   : ${dateBirth}\n` +
        `cep          : ${cep}\n` +
        `password     : ${password}\n`);
    
    return generatedId;
}


class Account {
    constructor(id, name, phoneNumber, dateBirth, cep) {
        this.id          = id;
        this.name        = name;
        this.phoneNumber = phoneNumber;
        this.dateBirth   = dateBirth;
        this.cep         = cep;
        // this.password    = password;
    }
    
    static login(name, password) {
        const data = getAccount(name, password);
        
        const accountId = data[0]
        const account   = data[1]
        
        if (account === null) {
            return null;
        }
        
        const _name       = account.name
        const phoneNumber = account.phoneNumber;
        const dateBirth   = account.dateBirth;
        const cep         = account.cep;
        
        return new Account(accountId, _name, phoneNumber, dateBirth, cep);
    }
    
    static register(name, phoneNumber, dateBirth, cep, password) {
        const accountId = addAccount(name, phoneNumber, dateBirth, cep, password);
        
        if (accountId === null) {
            return null;
        }
        
        return new Account(accountId, name, phoneNumber, dateBirth, cep);
    }
}

async function main() {
    let account;
    let choice;
    
    if (!fs.existsSync(DIRECTORY_DB)) {
        fs.mkdirSync(DIRECTORY_DB, { recursive: true });
    }
    
    [ACCOUNTS_DB, PURCHASES_DB].forEach((dbPath) => {
        if (!fs.existsSync(dbPath)) {
            fs.writeFileSync(dbPath, "");
        }
    });
    
    let loop = true;
    console.log("Menu de Opções:");
    
    while (loop) {
        console.log(
            "\n1. Login\n"  +
            "2. Registro\n" +
            "3. Sair\n");
        
        choice = await validInput("O que você deseja fazer? ");
        
        let name;
        let phoneNumber;
        let dateBirth;
        let cep;
        let password;
        
        switch (choice) {
            case "1":
                name     = await validInput("Digite seu nome: ");
                password = await validInput("Digite sua senha: ");
                
                account = Account.login(name, password);
                break;
            case "2":
                name        = await validInput("Digite seu nome: ");
                phoneNumber = await validInput("Digite seu número de telefone: ");
                dateBirth   = await validInput("Digite sua data de nascimento: ");
                cep         = await validInput("Digite seu CEP: ");
                password    = await validInput("Digite sua senha: ");
                
                if ("" in []) {
                    console.log("Você esqueceu de preencher um ou mais campos. Tente novamente!");
                    continue;
                }
                
                account = Account.register(name, phoneNumber, dateBirth, cep, password);
                break;
            case "3":
                console.log("Até logo!");
                process.exit();
        }
        
        
        if (account === null) {
            console.log(
                "Alguma coisa deu errado, tente novamente.\n" +
                
                "\n    Coisas que podem ter acontecido:\n" +
                "No login, você preencheu as credenciais incorretamente ou a conta não existe.\n" +
                "No registro, já existe uma conta com o mesmo nome e senha.");
            
            continue;
        }
        
        loop = false;
    }
    
    loop = true;
    console.log(`Olá, ${account.name}!`);
    
    while (loop) {
        console.log(
            "\n1. Exibir Catálogo\n"        +
            "2. Adicionar ao carrinho\n"    +
            "3. Remover do carrinho\n"      +
            "4. Mostrar carrinho\n"         +
            "5. Pagar produtos\n"           +
            "6. Ver histórico de compras\n" +
            "7. Sair\n");
        
        choice = await validInput("O que você deseja fazer? ");
        
        let index;
        let quantity;
        
        switch (choice) {
            case "1":
                console.log("Catálogo de Frutas:\n");
                printProducts();
                break;
            
            case "2":
                try {
                    index    = parseInt(await validInput("Qual produto você deseja adicionar ao carrinho? "));
                    quantity = parseInt(await validInput("Quantos do mesmo produto você deseja adicionar? "));
                    
                    if (quantity < 1) {
                        console.log("Você não pode adicionar menos que 1 produto no carrinho.");
                        continue;
                    }
                    
                    addToCart(index, quantity);
                    
                    console.log("O produto foi adicionado com sucesso ao carrinho.");
                } catch (e) {
                    console.log("Houve um erro ao adicionar o produto ao carrinho. O produto existe?");
                }
                break;
            
            case "3":
                try {
                    index    = parseInt(await validInput("Qual produto você deseja remover do carrinho? "));
                    quantity = parseInt(await validInput("Quantos do mesmo produto você deseja remover? "));
                    
                    if (quantity < 1) {
                        console.log("Você não pode remover menos que 1 produto no carrinho.");
                        continue;
                    }
                    
                    removeFromCart(index, quantity);
                    
                    console.log("O produto foi removido com sucesso do carrinho.");
                } catch (e) {
                    console.log("Houve um erro ao remover o produto do carrinho.");
                }
                break;
            
            case "4":
                console.log("Seu Carrinho:\n");
                showCart();
                break;
            
            case "5":
                payProducts(account.id);
                break;
            
            case "6":
                viewPurchases(account.id);
                break;
            
            case "7":
                console.log("\nAté logo!");
                loop = false;
                break;
            
            default:
                console.log("Opção inválida. Tente novamente!");
                break;
        }
    }
    
    process.exit();
}

if (process.mainModule.filename === __filename) {
    main();
}

File: combined.py
import os
import sys
import re
from random import randint

DIRECTORY_DB = './db/'
ACCOUNTS_DB  = './db/accounts.txt'
PURCHASES_DB = './db/purchases.txt'

products = {
    '1': {
        'name'        : 'Maçã',
        'price'       : 1.99,
        'description' : 'Vermelha/verde'
    },
    '2': {
        'name'        : 'Manga',
        'price'       : 5.50,
        'description' : 'Rosa/espada'
    },
    '3': {
        'name'        : 'Laranja',
        'price'       : 2.49,
        'description' : 'Lima/pera'
    },
    '4': {
        'name'        : 'Banana',
        'price'       : 0.99,
        'description' : 'Prata/nanica'
    },
    '5': {
        'name'        : 'Uva',
        'price'       : 2.90,
        'description' : 'Tompson/niagara'
    }
}

cart = {}


def parse_txt(filepath):
    data = {}
    last_id  = '-1'
    
    with open(filepath, 'r') as file:
        content = [item.strip() for item in file.readlines() if item.strip() != '']
    
    for line in content:
        match = re.match(r'([a-zA-Z0-9_\-]+)(?:\s+)?(?::)(?:\s+)?(.+)', line)
        
        if match is None:
            continue
        
        key_name  = match.group(1)
        key_value = match.group(2)
        
        if key_name == 'id':
            last_id = key_value
            data[key_value] = {}
        else:
            data[last_id][key_name] = key_value
    
    return data


def _get_random_id(limit=999_999_999_999):
    accounts = parse_txt(ACCOUNTS_DB)
    used_ids = []
    generated_id = None
    
    for account_id in accounts:
        used_ids.append(account_id)
    
    while True:
        generated_id = str(randint(1, limit))
        
        if generated_id not in used_ids:
            break
    
    return generated_id

def get_account(name, password):
    accounts = parse_txt(ACCOUNTS_DB)
    
    for account_id in accounts:
        account = accounts[account_id]
        
        if account['name'].lower() == name.lower() and account['password'] == password:
            return account_id, account
    
    return None, None

def add_account(name, phone_number, date_birth, cep, password):
    if get_account(name, password) != (None, None):
        return None
    
    generated_id = _get_random_id()
    
    with open(ACCOUNTS_DB, 'a') as file:
        file.write(
            f'id           : {generated_id}\n'
            f'name         : {name}\n'
            f'phone_number : {phone_number}\n'
            f'date_birth   : {date_birth}\n'
            f'cep          : {cep}\n'
            f'password     : {password}\n'
        )
    
    return generated_id


class Account:
    def __init__(self, _id, name, phone_number, date_birth, cep):
        self.id           = _id
        self.name         = name
        self.phone_number = phone_number
        self.date_birth   = date_birth
        self.cep          = cep
        # self.password     = password
    
    @classmethod
    def login(cls, name, password):
        account_id, account = get_account(name, password)
        
        if account is None:
            return None
        
        name         = account['name']
        phone_number = account['phone_number']
        date_birth   = account['date_birth']
        cep          = account['cep']
        
        return cls(account_id, name, phone_number, date_birth, cep)
    
    @classmethod
    def register(cls, name, phone_number, date_birth, cep, password):
        account_id = add_account(name, phone_number, date_birth, cep, password)
        
        if account_id is None:
            return None
        
        return cls(account_id, name, phone_number, date_birth, cep)


def _custom_price(price, dot=',', money='R$'):
    render = f'{price:.2f}'.replace('.', dot)
    return f'{money} {render}'


def print_products():
    for product_id in products:
        product = products[product_id]
        
        name = product['name']
        price = product['price']
        description = product['description']
        
        print(f'[{product_id}] {name}: {_custom_price(price)} - {description}')

def add_to_cart(index, quantity):
    product_id = str(index)
    
    if cart.get(product_id) is None:
        cart[product_id] = 0
    
    cart[product_id] += quantity

def remove_from_cart(index, quantity):
    product_id = str(index)
    
    cart[product_id] -= quantity
    
    if cart[product_id] < 1:
        del cart[product_id]

def show_cart():
    if len(cart) == 0:
        print('O carrinho está vazio.')
        return
    
    total = .0
    for product_id in cart:
        product = products[product_id]
        
        quantity    = cart[product_id]
        name        = product['name']
        price       = product['price'] * quantity
        description = product['description']
        
        total += price
        print(f'[{product_id}] {name} ({quantity}): {_custom_price(price)} - {description}')
    
    print(f'\nPreço Total: {_custom_price(total)}')

def pay_products(account_id):
    if len(cart) == 0:
        print('Não há nenhum produto em seu carrinho para ser pago.')
        return
    
    purchases = parse_txt(PURCHASES_DB)
    purchases[account_id] = purchases.get(account_id, {})
    
    total = .0
    for product_id in list(cart):
        if purchases[account_id].get(product_id) is None:
            purchases[account_id][product_id] = 0
        else:
            purchases[account_id][product_id] = int(purchases[account_id][product_id])
        
        purchases[account_id][product_id] += int(cart[product_id])
        
        total += products[product_id]['price'] * purchases[account_id][product_id]
        
        del cart[product_id]
    
    
    os.remove(PURCHASES_DB)
    
    with open(PURCHASES_DB, 'a') as file:
        for user_id in purchases:
            file.write(f'id : {user_id}\n')
            
            for product_id in purchases[user_id]:
                quantity = purchases[user_id][product_id]
                file.write(f'{product_id} : {quantity}\n')
    
    
    print(
        'Compra feita com sucesso!\n'
       f'Foi pago um total de {_custom_price(total)}')

def view_purchases(account_id):
    purchases = parse_txt(PURCHASES_DB)
    personal  = purchases.get(account_id, {})
    
    if len(personal) == 0:
        print('Não foi feita nenhuma compra nesta conta.')
        return
    
    total = .0
    for product_id in personal:
        product  = products[product_id]
        quantity = int(personal[product_id])
        
        name        = product['name']
        price       = product['price'] * quantity
        description = product['description']
        
        total += price
        print(f'[{product_id}] {name} ({quantity}): {_custom_price(price)} - {description}')
    
    print(f'\nTotal Pago: {_custom_price(total)}')


def main():
    # Cria os arquivos caso eles não existam
    for db_path in (ACCOUNTS_DB, PURCHASES_DB):
        if not os.path.exists(db_path):
            os.makedirs(DIRECTORY_DB, exist_ok=True)
            
            with open(db_path, 'w') as file:
                file.write('')
    
    
    print('Menu de Opções:')
    
    while True:
        print(
            '\n1. Login\n'
            '2. Registro\n'
            '3. Sair\n')
        
        choice = input('O que você deseja fazer? ').strip()
        
        if choice == '1':
            name     = input('Digite seu nome: ').strip()
            password = input('Digite sua senha: ').strip()
            
            account = Account.login(name, password)
        elif choice == '2':
            name         = input('Digite seu nome: ').strip()
            phone_number = input('Digite seu número de telefone: ').strip()
            date_birth   = input('Digite sua data de nascimento: ').strip()
            cep          = input('Digite seu CEP: ').strip()
            password     = input('Digite sua senha: ').strip()
            
            if '' in (name, phone_number, date_birth, cep, password):
                print('Você esqueceu de preencher um ou mais campos. Tente novamente!')
                continue
            
            account = Account.register(name, phone_number, date_birth, cep, password)
        elif choice == '3':
            print('Até logo!')
            sys.exit()
        else:
            print('Escolha inválida.')
            continue
        
        
        if account is None:
            print(
                'Alguma coisa deu errado, tente novamente.\n'
                
                '\n    Coisas que podem ter acontecido:\n'
                'No login, você preencheu as credenciais incorretamente ou a conta não existe.\n'
                'No registro, já existe uma conta com o mesmo nome e senha.')
            
            continue
        
        break
    
    
    print(f'Olá, {account.name}!')
    
    while True:
        print(
            '\n1. Exibir Catálogo\n'          
            '2. Adicionar ao carrinho\n'
            '3. Remover do carrinho\n'
            '4. Mostrar carrinho\n'
            '5. Pagar produtos\n'
            '6. Ver histórico de compras\n'
            '7. Sair\n')
        
        choice = input('O que você deseja fazer? ').strip()
        if choice == '1':
            print('Catálogo de Frutas:\n')
            print_products()
        
        elif choice == '2':
            try:
                index = int(input('Qual produto você deseja adicionar ao carrinho? ').strip())
                quantity = int(input('Quantos do mesmo produto você deseja adicionar? ').strip())
                
                if quantity < 1:
                    print('Você não pode adicionar menos que 1 produto no carrinho.')
                    continue
                
                add_to_cart(index, quantity)
                
                print('O produto foi adicionado com sucesso ao carrinho.')
            except IndexError:
                print('Houve um erro ao adicionar o produto ao carrinho. O produto existe?')
            except ValueError:
                print('O código do produto ou a quantidade não é um número.')
        
        elif choice == '3':
            try:
                index = int(input('Qual produto você deseja remover do carrinho? ').strip())
                quantity = int(input('Quantos do mesmo produto você deseja remover? ').strip())
                
                if quantity < 1:
                    print('Você não pode remover menos que 1 produto no carrinho.')
                    continue
                
                remove_from_cart(index, quantity)
                
                print('O produto foi removido com sucesso do carrinho.')
            except IndexError:
                print('Houve um erro ao remover o produto do carrinho. O produto não está no carrinho?')
            except ValueError:
                print('O código do produto ou a quantidade não é um número.')
        
        elif choice == '4':
            print('Seu Carrinho:\n')
            show_cart()
        
        elif choice == '5':
            pay_products(account.id)
        
        elif choice == '6':
            view_purchases(account.id)
        
        elif choice == '7':
            print("\nAté logo!")
            break
        
        else:
            print("Opção inválida. Tente novamente!")

if __name__ == '__main__':
    main()
